### PR TITLE
Hosted By traffic driver

### PR DESCRIPTION
**Pull request by lps88**
New 'thrasher' traffic driver for hosted campaigns. Can have 2, 3 or 4 cards.

To continue this work:
1. Check out `guardian/commercial-templates`
2. Edit `src/guardian-hosted/index.html` and `src/guardian-hosted/index.scss`
3. Run `npm run build`
4. Copy `build/guardian-hosted/index.html` and `build/guardian-hosted/index.css` in to the appropriate boxes in [the native template](https://www.google.com/dfp/59666047#delivery/CreateNativeAdFormatAndStyle/nativeStyleId=71215&creativeTemplateId=10062327) and save.
5. Click `Ad format details` to add/remove template variables if needed.
6. Open [localhost:9000/uk?adtest=lydia1](http://localhost:9000/uk?adtest=lydia1) and the thrasher should appear in the page.

Changes take a while to appear from DFP. To test style changes more quickly, copy the compiled css into the style tag in the iframe instead.

Any template changes may require updates to [this creative](https://www.google.com/dfp/59666047#delivery/PreviewCreative/orderId=341470047&lineItemId=131115447&creativeId=102066672447).

There is a field in the template called `numberOfElements` which determines how many cards are shown. It's currently set to two, but the creative has images etc set up for all four cards. Just change the `with-2-elements` class to quickly add/remove the 3rd and 4th.

upload_8_23_2016_at_4_51_15_pm
upload_8_23_2016_at_4_52_18_pm
upload_8_23_2016_at_4_53_05_pm

### CODE REVIEW

File: build/guardian-hosted/index.html [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<base target="_top">
+<div class="creative creative--hosted l-side-margins fc-container__inner"
+     data-link-name="Hosted Thrasher Multi">
+    <div class="creative--hosted__content">
+        <p class="fc-container__header__title">
+            from our advertisers
+        </p>
+        <div class="fc-container__body hosted-trafficdriver__body" data-link-name="">
+            <a class="hosted-trafficdriver__tile" href="[%link1%]" data-link-name="[%linkTracking1%]" style="
```
Comment (Calanthe) [inline build/guardian-hosted/index.html]: `__tile` or `__title`?

**Comment by regiskuckaertz** [comment]
Hey @GHaberis and @lps88 . I know there is an issue around breakpoints if the page has a scrollbar --- which is basically all the time --- because the view inside the iframe has different dimensions than its parent and thus all media queries are skewed. There are three solutions I can think of:

1- do nothing. this is a small gap and very unlikely to get noticed much anyway.
2- decrease breakpoints inside iframes. there probably is a sweet spot that could work across browsers and platforms, leaving a 0–5px gap globally; of course in the absence of scrollbar this would create the opposite problem.
3- create a routine that sends messages when a breakpoint changes. I'm currently rewriting the logic inside `detect` and this would be quite trivial, but at the cost of increasing the cyclomatic complexity of the CSS inside iframes.

Anyway, thought I'd let you know, don't spend time on this as I got it covered. Happy to discuss if you have another solution though.

**Comment by JonNorman** [comment]
Is this PR still going ahead?

**Comment by lps88** [comment]
@regiskuckaertz @GHaberis I haven't looked at this in ages, are you both happy for it to be merged?

**Comment by regiskuckaertz** [comment]
No 😃 

- Please take inspiration from other creatives (`index.js` and `index.scss`), notably the click tracking and the iframe ID
- Please do without the `grid-system` partial, we already have everything we need in core
- We don't track the `build` folder anymore, you can remove the generated assets

**Comment by lps88** [comment]
@regiskuckaertz Can I remove DevMode now? How do I get it to appear on production?

**Comment by regiskuckaertz** [comment]
@lps88 yes, this is all taken care of now, this var can go away. You should not have any problem seeing it on prod.

**Comment by lps88** [comment]
@regiskuckaertz I can't figure out work out why I can see it here:
http://m.code.dev-theguardian.com/uk/environment?adtest=lydia1
but not here:
https://www.theguardian.com/uk/environment?adtest=lydia1
Can you help?

**Comment by regiskuckaertz** [comment]
Weird. The creative was detected as not SSL compatible. I overrode the setting, let's see.

**Comment by regiskuckaertz** [comment]
@lps88 any update on this? Let me know when we can test.

File: src/guardian-hosted/web/index.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { getIframeId, getWebfonts, resizeIframeHeight, onViewport, reportClicks } from '../../_shared/js/messages';
+
+reportClicks();
+getIframeId()
+    .then(() => addColourContrastClass())
+    .then(() => getWebfonts())
+    .then(resizeIframeHeight)
```
Comment (regiskuckaertz) [inline src/guardian-hosted/web/index.js]: You don't need this line, `onViewport` will be called right away.

File: src/guardian-hosted/web/index.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    .then(function(){
+        onViewport(() => resizeIframeHeight());
+    });
+
+function addColourContrastClass () {
+    var div = document.getElementsByClassName('creative--hosted')[0];
+    var brandColour = div.getAttribute('data-brand-color');
+    if(isDark(brandColour)){
+        div.className += " hosted-bright";
```
Comment (regiskuckaertz) [inline src/guardian-hosted/web/index.js]: `div.classList.add('hosted--bright')`

Comment (regiskuckaertz) [inline src/guardian-hosted/web/index.js]: ... and you can even wrap it in a `write` call 😃

File: src/guardian-hosted/web/index.html [hunk 4]
```diff
@@ -0,0 +1,57 @@
+
+<div class="creative creative--hosted l-side-margins with-[%numberOfElements%]-elements"
```
Comment (regiskuckaertz) [inline src/guardian-hosted/web/index.html]: I think you can try replacing `l-side-margins` with `gs-container`. Let's see what it gives.

File: src/guardian-hosted/web/index.scss [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    @include mq(leftCol){
+        width: 140px;
+    }
+}
+
+@mixin side-margins-position($container-width) {
+    // stupid fix for sasslint AST error
+    $size: $container-width + 40px;
+
+    &:before,
```
Comment (regiskuckaertz) [inline src/guardian-hosted/web/index.scss]: for pseudo-elements, you need to use `::`, e.g. `::after`. One colon is for pseudo-_selectors_ and is deprecated for pseudo-elements.

File: src/guardian-hosted/web/index.scss [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    &:after {
+        width: calc((100% - #{$size}) / 2);
+    }
+}
+
+.l-side-margins {
```
Comment (regiskuckaertz) [inline src/guardian-hosted/web/index.scss]: hopefully with `gs-container` you can get rid of this one. Not a big deal though.

**Comment by lps88** [comment]
@regiskuckaertz thanks for the feedback! I've implemented those changes now, seems to all work fine.

**Comment by regiskuckaertz** [comment]
NB:

**Comment by regiskuckaertz** [comment]
Oops, sorry. I closed it by mistake 😅

**Pull request merged**